### PR TITLE
Profitbricks REST - Snapshot API

### PR DESCRIPTION
**Pull request by mirza-spc**
@jasminSPC

### CODE REVIEW

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/snapshot/UpdateSnapshotRequestBinder.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.Map;
+import org.apache.jclouds.profitbricks.rest.binder.BaseProfitBricksRequestBinder;
+import org.apache.jclouds.profitbricks.rest.domain.Snapshot;
+import org.jclouds.http.HttpRequest;
+import org.jclouds.json.Json;
+
+public class UpdateSnapshotRequestBinder extends BaseProfitBricksRequestBinder<Snapshot.Request.UpdatePayload> {
+
+   final Map<String, Object> requestBuilder;
+   final Json jsonBinder;
```
Comment (nacx) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/snapshot/UpdateSnapshotRequestBinder.java]: I've noticed these properties are used in all binders. Would it make sense to move them tot he parent class? Also, is there any reason why they're not private?

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import static org.testng.Assert.assertNotNull;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+@Test(groups = "live", testName = "SnapshotApiLiveTest")
+public class SnapshotApiLiveTest extends BaseProfitBricksLiveTest {
+   
+   DataCenter dataCenter;
+   Volume testVolume;
+   Snapshot testSnapshot;
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Make these variables private

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      assertNotNull(testVolume);
+      assertVolumeAvailable(testVolume);
+      
+      testSnapshot = api.volumeApi().createSnapshot(
+         Volume.Request.createSnapshotBuilder()
+            .dataCenterId(testVolume.dataCenterId())
+            .volumeId(testVolume.id())
+            .name("test-snapshot")
+            .description("snapshot desc...")
+            .build());
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Move this better to a `createSnapthot` test method.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      if (dataCenter != null)
+         deleteDataCenter(dataCenter.id());
+   }
+   
+   @Test
+   public void testList() {
+      List<Snapshot> snapshots = snapshotApi().getList();
+
+      assertNotNull(snapshots);
+      assertFalse(snapshots.isEmpty());
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Assert that it actually contains the just created snapshot?

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+   @Test
+   public void testGetSnapshot() {
+      Snapshot snapshot = snapshotApi().getSnapshot(testSnapshot.id());
+
+      assertNotNull(snapshot);
+      assertEquals(snapshot.id(), testSnapshot.id());
+      assertEquals(snapshot.properties().name(), "test-snapshot");
+   }  
+   
+   @Test(dependsOnMethods = "testGetSnapshot")
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Consider changing this to depend on the create snapshot test once it exists.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      snapshotApi().deleteSnapshot(testSnapshot.id());
+      assertSnapshotRemoved(testSnapshot);
+   }
+   
+   private SnapshotApi snapshotApi() {
+      return api.snapshotApi();
+   }
+   
+   private void assertVolumeAvailable(Volume volume) {
+      assertRandom(new Predicate<VolumeRef>() {
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Just seen this... Would it be more convenient for this method to be named `assertPredicate`? The current name seems to mean that it will perform arbitrary assertions...

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "PATCH", "/rest/snapshots/some-id", "{\"name\": \"new-snapshot-name\", \"description\": \"description...\"}");
+   }
+   
+   @Test
+   public void testDelete() throws InterruptedException {
+      server.enqueue(
+         new MockResponse().setBody("")
+      );
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java]: Instead of an empty body, wouldn't it be better to configure a response with the 204 status code?

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+   
+   
+   @Test
+   public void testGetListWith404() throws InterruptedException {
+      server.enqueue(response404());
+      List<Snapshot> list = snapshotApi().getList(new DepthOptions().depth(1));
+      assertTrue(list.isEmpty());
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "GET", "/snapshots?depth=1");
+   }
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java]: Add a test for the 404 fallback for the other `getList` too.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+   public void testGetSnapshotWith404() throws InterruptedException {
+      server.enqueue(response404());
+
+      Snapshot snapshot = snapshotApi().getSnapshot("some-id");
+      
+      assertEquals(snapshot, null);
+
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "GET", "/snapshots/some-id");
+   }
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiMockTest.java]: Add a test for the 404 fallback for the other `getSnapshot` too.

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java [hunk 10]
```diff
@@ -19,29 +19,72 @@
...
 
 @Path("/snapshots")
 @RequestFilters(BasicAuthentication.class)
 public interface SnapshotApi extends Closeable {
    
+   @Named("snapshot:list")
+   @GET
+   @SelectJson("items")
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<Snapshot> getList();
```
Comment (nacx) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java]: I've just noticed this pattern. In other jclouds APIs, the common name for list methods is just `list` (instead of getList). Mind doing the change?

**Comment by nacx** [comment]
Thanks @mirza-spc! This looks pretty good; all comments are minors.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java (hunk 3 above)
Comment (mirza-spc) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: That is already tested in volume api, there is not create snapshot in snapshot api as per docs.

Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Oh, you're right. Dismiss this comment then! :)

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java (hunk 10 above)
Comment (mirza-spc) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java]: ok but you still have getSnapshot, should it be just get? or createSnapshot => create? that would seem along the lines of such a convention. or is it just that the list method is a special case?

Comment (nacx) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java]: Yes, that's the convention we try to use now, so it applies to all mehtods.

Comment (mirza-spc) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java]: this should mean that all of the methods of all of the apis in profitbricks rest should be updated to the convention format?

Comment (nacx) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApi.java]: Yes, but let's leave that for another PR and just amend the ones in the snapshot api in this one.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+      assertFalse(snapshots.isEmpty());
+      
+      Snapshot newSnapshot = null;
+      
+      for (Snapshot snapshot : snapshots) {
+         if (snapshot.id().equals(testSnapshot.id())) {
+            newSnapshot = snapshot;
+            break;
+         }
+      }
```
Comment (nacx) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/SnapshotApiLiveTest.java]: Better use:
```java
assertTrue(Iterables.any(snapshots, new Predicate<Snapshot>() {
   @Override public boolean apply(Snapshot input) {
      return input.id().equals(testSnapshot.id());
   }
}));
```

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/snapshot/UpdateSnapshotRequestBinder.java (hunk 1 above)
Comment (nacx) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/snapshot/UpdateSnapshotRequestBinder.java]: What about this comment? Something that makes sense for a separate PR? Also I've noticed some binders declare the variables in the default scope. If there is no good reason for that, they should be made private. We could address both in a new PR.

Comment (mirza-spc) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/snapshot/UpdateSnapshotRequestBinder.java]: I've refactored it out into the base class. Good thing for noticing it. I think they are only two binders that build the form encoded data, all others are using json.

**Comment by nacx** [comment]
Thanks @mirza-spc! Mind squashing the commits to prepare the merge?

**Comment by nacx** [comment]
Pushed to master as [ca57d90d](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/ca57d90d). Thanks!
